### PR TITLE
fix: switch DnD to modern context API (was legacy)

### PR DESCRIPTION
**Pull request by ehahn9**
DnD addon was relying on legacy context api which finally broke with React 17 (couldn't resize events, etc.).

Fixes #1795. Related to #1776.

### CODE REVIEW

**Comment by ehahn9** [comment]
Hmmm... might still have remaining v17 compatibility issues due to DnD's use of findDOMNode (deprecated). Working on that too.

**Comment by jquense** [comment]
let me know when you want to merge 👍

**Comment by ehahn9** [comment]
Thanks, @jquense -- the context switch was easy. The use of deprecated findDOMNode is harder (am working on that now).

The original DnD addon is looking like it need some serious love (very non-DRY, odd js idioms throughout) -- I might take a crack some rejuvenation throughout, but I also want to see it working with React 17 asap.

I'll ping you when I have something useful to review!

**Comment by ehahn9** [comment]
Hi @jquense... I just realized that the [examples build](http://jquense.github.io/react-big-calendar/examples/) on github seem to be of a different vintage than those on master? (for example, the page's dropdown is missing more recently-added examples).

I've been working on finishing up the DnD refresh and thought I had introduced a regression but noticed that the examples built from master have the same probs (!).

I'll dive into this but wondering if you had thoughts on why the examples page is out-of-date? (or maybe I'm missing something...)

**Comment by jquense** [comment]
@ehahn9 i just always forget to deploy new docs which is why they are out of date. we should make that automatic with the version release tho

**Comment by ehahn9** [comment]
HI @jquense  - I [finally, sorry] had a chance to work on this. Many changes:

1. Use modern react context API (was legacy)
2. Uses refs rather than `findDOMNode` throughout the addin (although alas there is a bunch in the main rbc code, thus preventing react strict mode...)
3. Moved the DnD readme to the top-level of the addin (was buried in code comment)
4. Storybook: updated to v6, respect allDay in events
5. Fixed event propagation bugs (seen only in react 17)
6. Fixed a number of bounding box and fencepost errors in WeekView
7. Changed the HOC to extend, rather than repeat `propTypes` and `defaultProps`

I had hoped to do a LOT more (css modules for scoped addon classnames, and major cleanup of original implementation which is difficult to maintain) -- maybe I'll get back to that at some point!

I'm still a newbie, so apologies if this PR needs more work (happy to try!).

**Comment by yann-combarnous** [comment]
@jquense, any chance you could review and possibly merge this PR? With DND resize broken in react 17, would be great to have it fixed.

**Comment by ehahn9** [comment]
Hi @jquense  - oh apologies - I'm new to this and I wasn't even aware that this was waiting on me. Sorry for asking for a lesson in OSS, but what's the protocol when one is reviewing their own PR? Do I just click squash + merge or should others be reviewing first/also? thx.

**Comment by jquense** [comment]
@ehahn9 generally you still want someone else to review it, but i've not been prompt on that sorry!

**Comment by jquense** [comment]
There is a lot of different fixes and changes here that make it a bit hard to review all at once. Generally its better to split these things into separate PR's, e.g. one for event propagation bugs, one for context changes, etc.

File: examples/demos/dnd.js [hunk 1]
```diff
@@ -55,7 +55,7 @@ class Dnd extends React.Component {
 
     const nextEvents = events.map(existingEvent => {
       return existingEvent.id == event.id
-        ? { ...existingEvent, start, end }
+        ? { ...existingEvent, start, end, allDay }
```
Comment (jquense) [inline examples/demos/dnd.js:58]: doesn't the `existingEvent` already have this?

File: src/addons/dragAndDrop/EventContainerWrapper.js [hunk 2]
```diff
@@ -73,43 +59,31 @@ class EventContainerWrapper extends React.Component {
...
-    let end = dates.add(
-      currentSlot,
-      dates.diff(eventStart, eventEnd, 'minutes'),
-      'minutes'
-    )
-
-    this.update(event, slotMetrics.getRange(currentSlot, end, false, true))
+    const { duration } = eventTimes(event, accessors)
+    let newEnd = dates.add(newSlot, duration, 'milliseconds')
+    this.update(event, slotMetrics.getRange(newSlot, newEnd, false, true))
```
Comment (jquense) [inline src/addons/dragAndDrop/EventContainerWrapper.js:74]: I'm not sure i understand what the changes in this file are fixing? can you walk me through it?

File: src/addons/dragAndDrop/EventContainerWrapper.js [hunk 3]
```diff
@@ -271,8 +246,10 @@ class EventContainerWrapper extends React.Component {
       ),
     })
   }
-}
 
-EventContainerWrapper.propTypes = propTypes
+  render() {
+    return <div ref={this.ref}>{this.renderContent()}</div>
```
Comment (jquense) [inline src/addons/dragAndDrop/EventContainerWrapper.js:251]: i don't think spliting renderContent out as a separate function adds much? maybe just have `render()`.

On the another note, I think it might more more sense to pass the ref to `children` instead of adding a new DOM element

File: src/addons/dragAndDrop/EventWrapper.js [hunk 4]
```diff
@@ -32,28 +23,28 @@ class EventWrapper extends React.Component {
...
     this.context.draggable.onBeginAction(this.props.event, 'resize', 'UP')
   }
   handleResizeDown = e => {
     if (e.button !== 0) return
-    e.stopPropagation()
     this.context.draggable.onBeginAction(this.props.event, 'resize', 'DOWN')
   }
   handleResizeLeft = e => {
     if (e.button !== 0) return
-    e.stopPropagation()
```
Comment (jquense) [inline src/addons/dragAndDrop/EventWrapper.js:45]: It seems like removing these would create bugs, not fix them? Can you walk me through why this is necessary?

File: src/addons/dragAndDrop/WeekWrapper.js [hunk 5]
```diff
@@ -79,42 +61,34 @@ class WeekWrapper extends React.Component {
     this.setState({ segment })
   }
 
-  handleMove = ({ x, y }, node, draggedEvent) => {
+  handleMove = (point, bounds, draggedEvent) => {
```
Comment (jquense) [inline src/addons/dragAndDrop/WeekWrapper.js:64]: All the changes in this file are hard to follow (though i'm sure correct and needed). What is is going on here, and how should test that it works?

**Review by jquense — COMMENTED** [review]
lots of great work, sorry i'm having a hard time following it!

File: examples/demos/dnd.js (hunk 1 above)
Comment (ehahn9) [reply examples/demos/dnd.js:58]: The old story's `moveEvent` handler did't honor changes affecting `allDay`. This change purports to make it possible to demo moving events in and out of the allDay gutter - supported by DnD but previously not supported by the story as it ignored the new value for allDay.

File: src/addons/dragAndDrop/EventContainerWrapper.js (hunk 3 above)
Comment (ehahn9) [reply src/addons/dragAndDrop/EventContainerWrapper.js:251]: Absolutely right. I was having trouble forwarding the ref and took the cowards way out. I'll try to work on this.

File: src/addons/dragAndDrop/EventWrapper.js (hunk 4 above)
Comment (ehahn9) [reply src/addons/dragAndDrop/EventWrapper.js:45]: Ok, this one is tricky. the Dnd addon uses rbc's selectable() mechanism to monitor mouse events before/during drags. selectable needs to see those events in order to work. The event wrapper stopped the propagation and so the selection logic was basically broken.

So first, I removed the stop propagation - that helped a lot.

But because the resize anchors are on top of the drag (move) anchor, clicking on a resize would also start a drag/move, which would abort the resize, etc.

So I added a check (see :42) to ignore this case:

```
    // hack: because of the way the anchors are arranged in the DOM, resize
    // anchor events will bubble up to the move anchor listener. Don't start
    // move operations when we're on a resize anchor.
    const isResizeHandle = e.target.className.includes('rbc-addons-dnd-resize')
    if (!isResizeHandle)
      this.context.draggable.onBeginAction(this.props.event, 'move')
```

Now the million dollar question is _why did this ever work?_ My only theory is that react 17 changed where the event listeners were placed and _maybe_ that subtle changed the even propagation?

I would love to back all of this out, but it seemed to be required both in reading the code and trying it.

File: src/addons/dragAndDrop/WeekWrapper.js (hunk 5 above)
Comment (ehahn9) [reply src/addons/dragAndDrop/WeekWrapper.js:64]: Purely idiomatic: all the other event handlers had the `(point, bounds)` signature (as does the similar code in `EventContainerWrapper`) so I was just conforming the idiom.

BTW, if you look closely, you'll see that `WeekWrapper` and `EventContainerWrapper` are in DIRE need of being DRY'ed out -- they basically do the same exact thing, with the main difference being the orientation of the event (horizontal or vertical). I spent a few days on that, but had to abandon due to time.

File: src/addons/dragAndDrop/EventContainerWrapper.js (hunk 2 above)
Comment (ehahn9) [reply src/addons/dragAndDrop/EventContainerWrapper.js:74]: Good catch. The problem I was addressing was that the addon would sometimes fail to update the event `end` on a move. This change always sets `end = start + duration` where start = `newSlot` (e.g. the target time slot at the end of the move).

File: examples/demos/dnd.js (hunk 1 above)
Comment (ehahn9) [reply examples/demos/dnd.js:58]: I should have done this as a separate PR for sure! Newbie/learning.

**Comment by ehahn9** [comment]
I'll wait to hear back from you (@jquense)  before working further on this PR. Agreed, I really should have done this in more atomic pieces (and will next time!). Alas, I find the DnD addon to be super hard to work on (maybe even a rewrite) and so my energy was flagging during this... apologies.

File: src/addons/dragAndDrop/WeekWrapper.js (hunk 5 above)
Comment (ehahn9) [reply src/addons/dragAndDrop/WeekWrapper.js:64]: ... oh, and the other changes (sorry, that's probably what you were asking about) - were all about making the date handling correct during drags - which had a number of probs, some of which might have been introduce with the (required) upgrade to date-arithmetic (which was out-of-date and causing some probs).

Happy to answer in more detail about these edits although my memory is fading a bit on this.

Comment (ehahn9) [reply src/addons/dragAndDrop/WeekWrapper.js:64]: testing: I think the only way is via storybook/examples -- the code seems too intertwined IMHO for proper testing (definitely a good should a rewrite occur!)

File: src/addons/dragAndDrop/EventWrapper.js (hunk 4 above)
Comment (jquense) [reply src/addons/dragAndDrop/EventWrapper.js:45]: sounds like you got it all handled 👍

File: src/addons/dragAndDrop/WeekWrapper.js (hunk 5 above)
Comment (jquense) [reply src/addons/dragAndDrop/WeekWrapper.js:64]: fine to defer the testing if there isn't a good way to write a unit test. I was thinking of the common-ish case where we fix one of these date bounds issues and introduce another one, partly because the calculations have no tests asserting their current behavior so easy to break stuff!

**Review by jquense — APPROVED** [review]
I would rename the PR so semantic release doesn't call all of this a "fix", might be worth hand rebasing the commits into different messages and then doing a "merge" instead of a "squash" but up to you.

**Comment by ehahn9** [comment]
Thanks @jquense  - just to be sure I'm doing the right thing: are you suggesting I replace this PR with smaller, more focussed ones? (I presume there's a way to cancel this PR?) I'm happy to work on that! I know there's some time pressure for this due to v17 so I hope I'm not holding things up.

**Comment by jquense** [comment]
>  just to be sure I'm doing the right thing: are you suggesting I replace this PR with smaller, more focussed ones? (I presume there's a way to cancel this PR?)

not quite, the suggestion was to rebase the commits in the current PR to something like
- fix: event propagation bugs
- fix: drag and drop bounds issues
- fix: update to modern context

etc, etc, and then merging instead of squash and merge so the commits stay intact on master. That said it all involves a bunch of Git-fu, that you may or may not be comfortable with, so fine to merge as is if you want

**Comment by ehahn9** [comment]
Just rebased all the commits and reworded - lmk if this looks okay to you -- then I'm happy to merge (merge commit, no squash)

**Comment by jquense** [comment]
very nice 👍

**Comment by jquense** [comment]
🎉  very nice

**Comment by nlevchuk** [comment]
@ehahn9 @jquense Thank you guys for your work

**Comment by yann-combarnous** [comment]
@ehahn9 , @jquense , thx for this! Any chance dnd will be made to work wit touch events at some stage? 

Dragging works but the initial event selection is very hard to get to work, takes generally a few seconds to manage selection.

**Comment by ehahn9** [comment]
@yann-combarnous - although more knowledgeable folks might disagree, I suspect the addon might be at the limits of maintainability at this point :-(. 

ps: I took a stab at [rewriting it from scratch as a completely separate project using react-dnd,](https://github.com/ehahn9/react-big-calendar-dnd) but ran into some architectural problems with resize animation.
